### PR TITLE
Add approaches for two-fer

### PR DESCRIPTION
**Pull request by IsaacG**

### CODE REVIEW

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    name="$1"
+fi
+
+echo "One for ${name}, one for me."
+```
+
+## Shorthand testing
+
+For simple conditionals, the `if` syntax is not needed.
+This is similary to the ternary operator in some ways.
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:33]:
```suggestion
This is similar to the ternary operator in some ways.
```

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+## Using printf
+
+The `printf` built-in can be used for string formatting.
+It is more flexible than `echo` and is safer for handling arbitrary inputs.
+(Try printing out the string `"-n"` with `echo`.)
+However, when using `printf`, one does need to remember to append a newline!
+
+```bash
+printf 'One for %s, one for me." "${1-you}"
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:48]:
```suggestion
printf 'One for %s, one for me.' "${1-you}"
```

Comment (glennj) [reply exercises/practice/two-fer/.approaches/introduction.md:48]: And did you forget to append a newline?

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+# Empty strings
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:51]: https://exercism.org/docs/building/markdown/markdown

Only 1 level-1 heading per file.
```suggestion
## Empty strings
```

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+On the one hand, the empty string may not be a valid name and the default "you" might be preferable.
+On the other hand, maybe the empty string is intended.
+Additionally, the input `" "` is generally accepted as a valid input.
+
+The empty string can be replaced with the default value, using tests or parameter expansiions.
+
+```bash
+[[ -z "$1" ]] && name=you || name="$1"
+printf 'One for %s, one for me." "${name}"
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:63]:
```suggestion
printf 'One for %s, one for me.\n' "${name}"
```

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+Paramter expansions also support testing for "unset or null" by adding a `:` to the syntax; without the `:`, only an unset variable is replaced with the default value.
+
+```bash
+printf 'One for %s, one for me." "${1:-you}"
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:69]:
```suggestion
printf 'One for %s, one for me.\n' "${1:-you}"
```

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+## POSIX tests
+
+The arithmetic test, `((` and the string test, `[[` are bashisms and not POSIX sh.
+The POSIX test is then `[` command.
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:75]:
```suggestion
The POSIX test is the `[` command.
```

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+When writing bash scripts, the built-in `((` and `[[` should be prefered; they are faster as they are built-ins (and not commands) and bash does not perform word splitting on them.
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:76]: Splitting the long sentence.
```suggestion
When writing bash scripts, the built-in [`((`][arithmetic] and [`[[`][conditional] should be prefered.
They are faster as they are built-ins (and not commands).
Additionally, variables inside `[[...]]` can be unquoted: bash does not perform word splitting or filename expansion inside the conditional construct.

[conditional]: https://wiki.bash-hackers.org/syntax/ccmd/conditional_expression
[arithmetic]: https://wiki.bash-hackers.org/syntax/ccmd/arithmetic_eval
```
With the (hard to explain briefly) proviso about quoting/unquoting the pattern on the right-hand side of `==` `!=` and `=~`

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+When using the `[` command, it is crucial that variable expansions be quoted.
+
+```bash
+[ "$#" -eq 0 ] && name=you || name="$1"
+printf 'One for %s, one for me." "${name}"
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:81]:
```suggestion
printf 'One for %s, one for me.\n' "${name}"
```
Same cut-and-paste error.

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+## References
+
+[default-value]: https://wiki.bash-hackers.org/syntax/pe#use_a_default_value
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:86]: If there's no text after the header, remove it

**Review by glennj — CHANGES_REQUESTED** [review]
Really nice work!

File: exercises/practice/two-fer/.approaches/introduction.md (hunk 9 above)
Comment (IsaacG) [reply exercises/practice/two-fer/.approaches/introduction.md:91]: Dropped.

File: exercises/practice/two-fer/.approaches/introduction.md (hunk 2 above)
Comment (IsaacG) [reply exercises/practice/two-fer/.approaches/introduction.md]: Hah. Yes. Pretty ironic :)

**Comment by IsaacG** [comment]
> Really nice work!

Thank you! And thanks for catching all my silly spelling mistakes!

File: exercises/practice/two-fer/.approaches/introduction.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+## POSIX tests
+
+The arithmetic test, `((` and the string test, `[[` are bashisms and not POSIX sh.
+The POSIX test is the `[` command.
+When writing bash scripts, the built-in [`((`][arithmetic] and [`[[`][conditional] should be prefered.
+They are faster as they are built-ins (and not commands).
+Additionally, variables inside `[[...]]` can be unquoted: bash does not perform word splitting or filename expansion inside the conditional construct.
+
+[conditional]: https://wiki.bash-hackers.org/syntax/ccmd/conditional_expression
+[arithmetic]: https://wiki.bash-hackers.org/syntax/ccmd/arithmetic_eval
```
Comment (glennj) [inline exercises/practice/two-fer/.approaches/introduction.md:81]: Minor nitpick: these could be placed at the bottom of the file.

**Review by glennj — APPROVED** [review]

File: exercises/practice/two-fer/.approaches/introduction.md (hunk 10 above)
Comment (kotp) [reply exercises/practice/two-fer/.approaches/introduction.md:81]: Preferred, probably, to do so.  I had to look twice thinking they were missing.

**Review by kotp — APPROVED** [review]

File: exercises/practice/two-fer/.approaches/introduction.md (hunk 10 above)
Comment (IsaacG) [reply exercises/practice/two-fer/.approaches/introduction.md]: My bad. Moved.
